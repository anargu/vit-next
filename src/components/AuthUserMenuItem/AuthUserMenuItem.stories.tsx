import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { AuthUserMenuItem } from "./AuthUserMenuItem";

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'UI/AuthUserMenuItem',
  component: AuthUserMenuItem,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
  },
} as ComponentMeta<typeof AuthUserMenuItem>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof AuthUserMenuItem> = (args : any) => (
  <div className="bg-gray-100"><AuthUserMenuItem {...args} /></div>
);

export const Default = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Default.args = {
};


