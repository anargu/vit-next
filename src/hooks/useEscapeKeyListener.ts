import { useEffect } from 'react';

const useKeyUpListener = (keyString : string, callback: (arg0: KeyboardEvent) => void) => {
  useEffect(() => {
    const handleKeyUp = (event : KeyboardEvent) => {
      if (event.key === keyString) {
        callback(event);
      }
    };

    document.addEventListener('keyup', handleKeyUp);

    return () => {
      document.removeEventListener('keyup', handleKeyUp);
    };
  }, [callback]);
};

export default useKeyUpListener;

