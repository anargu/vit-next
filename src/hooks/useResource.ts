import { showDefaultNotification } from "../components/Notification/Notification";
import { Resource, VITResource } from "../core/entities";

export const useResource = () => {

  const share = (resource : Resource | VITResource) => {
    navigator.clipboard.writeText(resource?.url || "");

    showDefaultNotification("Link copied to Clipboard.");
  };

  return { share };
};
