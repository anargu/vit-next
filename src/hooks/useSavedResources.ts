import { useEffect, useState } from "react";
import { SAVED_LINK_KEY } from "../components/ResourceCard/ResourceCard";
import { VITResource } from "../core/entities";

export const useSavedResources = () => {

  const [savedResources, setSavedResources] = useState<null | VITResource[]>(null);

  useEffect(() => {
    if (!!savedResources) return;

    const rawData = localStorage.getItem(SAVED_LINK_KEY);
    if (!rawData) return setSavedResources([]);
    
    const savedPosts : VITResource[] | null = JSON.parse(rawData);
    if (!savedPosts) return setSavedResources([]);

    /* const _resources = savedPosts.map((post) => Resource.fromVITResource(post)); */
    
    setSavedResources(savedPosts);
  }, [savedResources]);

  return { savedResources };
};
