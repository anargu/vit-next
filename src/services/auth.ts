import { signInWithRedirect, GoogleAuthProvider } from "firebase/auth";
import { doc, getDoc, setDoc, updateDoc } from "firebase/firestore";
import { auth, firestore, getRedirectResult } from "../firebase"

export type AuthenticatedUser = {
  id: string,
  email: string | null,
  displayName: string | null,
}

const googleProvider = new GoogleAuthProvider();

const signIn = async () => {
  await signInWithRedirect(auth, googleProvider)
};

const getSignInResult = async () => {
  return new Promise<AuthenticatedUser | null>((resolve, reject) => {
    getRedirectResult(auth)
      .then((result) => {
        const user = result?.user;

        if (!user) return resolve(null);

        resolve({
          id: user.uid,
          email: user.email,
          displayName: user.displayName,
        });
      }).catch((error) => {
        /* const errorCode = error.code; */
        /* const errorMessage = error.message; */
        /* // The email of the user's account used. */
        /* const email = error.customData.email; */
        console.error(`>>> Error: ${error.message}.\nCode: ${error.code}.\nUser:${error.customData?.email}`);

        reject(error);
      });
  });
};

const upsertUser = async (user : AuthenticatedUser) => {
  try {
    const docRef = doc(firestore, `users/${user.id}`);

    const userSnap = await getDoc(docRef);

    const upsertUserPromise = userSnap.exists()
      ? updateDoc(userSnap.ref, { lastSignIn: new Date() })
      : setDoc(userSnap.ref, {
          id: user.id,
          updatedAt: new Date(),
          createdAt: new Date(),
          lastSignIn: new Date(),
          email: user.email,
          displayName: user.displayName,
        });

    await upsertUserPromise;
  } catch (error) {
    console.error(`Error at upsertUser: ${error}`);
  }
};

export {
  signIn,
  getSignInResult,
  upsertUser,
};

