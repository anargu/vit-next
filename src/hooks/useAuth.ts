import { onAuthStateChanged } from "firebase/auth";
import { useEffect, useState } from "react";
import { User } from "../core/entities";
import { auth } from "../firebase";

// Hook
export function useAuth () {
  const [isAuthenticated, setIsAuthenticated] = useState(false);
  const [authUser, setAuthUser] = useState<User | null>(null);

  useEffect(() => {
    if (isAuthenticated) return;

    const unSub = onAuthStateChanged(auth, (firebaseUser) => {
      if (!firebaseUser) return;

      const authUser : User = {
        id: firebaseUser!.uid,
        email: firebaseUser!.email!,
        displayName: firebaseUser!.displayName!,
      };

      setIsAuthenticated(!!authUser);
      setAuthUser(authUser);
    });

    return () => unSub && unSub();
  }, [isAuthenticated]);

  return {
    isAuthenticated,
    authUser,
  };
}


