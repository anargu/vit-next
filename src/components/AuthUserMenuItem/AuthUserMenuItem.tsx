import { useAuth } from "@/src/hooks/useAuth";
import { logOut } from "@/src/services/auth";
import { useRouter } from "next/router";
import { useMemo, useState } from "react";
import User from "../../../public/assets/user.svg";
import { CircularLoader } from "../Loaders";
import { NavBarItemMenu } from "../NavBar/NavBarMenu";

export const AuthUserMenuItem = () => {
  const router = useRouter();
  const { isAuthenticated, authUser } = useAuth();
  const [ isLoading, setIsLoading ] = useState(false);

  const logOut_ = async () => {
    setIsLoading(true);

    await logOut();

    // redirects
    router.push("/login");

    setIsLoading(false);
  };

  const menu = useMemo(() => (
    isAuthenticated
      ? [
        { label: `Hi ${authUser?.email}` },
        { label: "Log out", onClick: logOut_ },
      ]
      : [
        { label: "Sign In", url: "/login" },
      ]
  ), [isAuthenticated]) ; 

  return (
    <NavBarItemMenu
      key={`auth-user-${isAuthenticated}`}
      menu={menu}
      right
    >
      <span className="cursor-pointer inline-flex items-center gap-2">
        <User />
        {isLoading
          ? (
            <CircularLoader colorClassname="text-stone-800" />
          )
          : null}
      </span>
    </NavBarItemMenu>
  );
};

