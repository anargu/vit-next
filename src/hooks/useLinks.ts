import { useEffect, useMemo, useState } from "react";
import useLocalStorageState from "use-local-storage-state";

import { useAuth } from "./useAuth";
import { DOMAIN_REGEX, SAVED_LINK_KEY } from "../core/constants";
import { fetchMetadataFromURL } from "../services/metadataService";
import { JSONResource, LinkPrivacy, Resource } from "../core/entities";
import { FirestoreStorageStrategy } from "../services/FirestoreStorageStrategy";
import { ga, ON_CLICK_SAVE_POST_EVENT, SAVE_POST_SUCESSFULLY_EVENT } from "../../lib/ga";

const firestoreDataManager = new FirestoreStorageStrategy();

export type UpdatePrivacySettingFn = (resourceId : string, privacy : LinkPrivacy) => Promise<void>;
export type SaveResourceFn = (resource : string | Resource) => Promise<void>;
export type DeleteResourceFn = (resourceId : string) => Promise<void>;

export interface useLinksOptions {
  listenSavedLinks?: boolean;
  listenFeedLinks?: boolean;
};

export function useLinks (opts: useLinksOptions = { listenSavedLinks: false, listenFeedLinks: false }) {

  const { authUser, isAuthenticated } = useAuth();

  const [_userLinks, _setUserLinks] = useState<Resource[] | null>([]);

  const [feedLinks, setFeedLinks] = useState<Resource[] | null>([]);

  // localLinks are saved links in LocalStorage 
  const [localLinks, setLocalLinks, { removeItem: deleteLocalLinks }] = useLocalStorageState<JSONResource[]>(SAVED_LINK_KEY, { defaultValue: [] });

  const userLinks = useMemo(() => {
    if (isAuthenticated) return _userLinks;

    return localLinks.map((localLink) => Resource.fromJSONResource(localLink));
  }, [isAuthenticated, localLinks, _userLinks]);

  // TODO: Implement this
  /* const initListeningLinksByUserId = (userId: string) => { */
  /*   const unSub = firestoreDataManager.listenLinksbyUserId(userId, (resources) => { */
  /*     _setUserLinks(resources); */
  /*   }); */
  /**/
  /*   return unSub; */
  /* }; */

  useEffect(() => {
    if (!authUser) return;
    if (!opts.listenSavedLinks) return;

    const unSub = firestoreDataManager.listenLinksbyUserId(authUser!.id, (resources) => {
      _setUserLinks(resources);
    });

    return () => {
      unSub?.();
    };
  }, [authUser?.id, opts.listenSavedLinks]);

  useEffect(() => {
    if (!opts.listenFeedLinks) return;

    const unSub = firestoreDataManager.listenFeedLinks((resources) => {
      setFeedLinks(resources);
    });

    return () => {
      unSub?.();
    };
  }, [opts.listenFeedLinks]);

  const insertLink = async (url : string) => {
    try {
      ga.event({ action: ON_CLICK_SAVE_POST_EVENT, params: {} });

      // Fetch Metadata of URL string
      let resourceMetadata = await fetchMetadataFromURL(url);

      // Save resource and its metadata
      const insertPromise = isAuthenticated
        ? firestoreDataManager.insert(resourceMetadata, authUser)
        : setLocalLinks([...localLinks, resourceMetadata.toJSON()]);

      await insertPromise;

      // Track analytics event
      const domain = url.match(DOMAIN_REGEX)?.[0];

      ga.event({
        action: SAVE_POST_SUCESSFULLY_EVENT,
        params: { domain: domain }
      });
    } catch (error : any) {
      console.error(error);

      throw error;
    }
  };

  const updateLinkPrivacy = async (resourceId: string, privacy: LinkPrivacy) => {
    try {
      if (!isAuthenticated) {
        throw Error("Operation cannot be done if user is not logged in.");
      }

      // User is logged in then update to DB.
      await firestoreDataManager.updateLinkPrivacy(resourceId, privacy);

    } catch (error : any) {
      console.error(error);

      throw error;
    }
  };

  const deleteLink = async (resourceId: string) => {
    try {
      if (!isAuthenticated) {
        setLocalLinks(localLinks.filter((_link) => _link.id !== resourceId));
        return;
      }

      await firestoreDataManager.delete(resourceId);
    } catch (error : any) {
      console.error(error);

      throw error;
    }
  };

  return {
    localLinks,
    deleteLocalLinks,

    userLinks,
    feedLinks,

    insertLink,
    deleteLink,
    updateLinkPrivacy,

    bulkInsertLinks: firestoreDataManager.bulkInsert,
  };
}
