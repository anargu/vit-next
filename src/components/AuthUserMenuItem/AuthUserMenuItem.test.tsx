import { render } from "@testing-library/react";
import { useAuth } from "@/src/hooks/useAuth";
import { AuthUserMenuItem } from "./AuthUserMenuItem";
import { AuthenticatedUser } from "@/src/services/auth";

jest.mock("@/src/hooks/useAuth", () => {
  return ({
    useAuth: jest.fn(() => ({
      isAuthenticated: false,
      authUser: null,
    })),
  });
});

describe("AuthUserMenuItem", () => {
  it("renders AuthUserMenuItem", () => {
    const { container } = render(<AuthUserMenuItem />);

    expect(container).toMatchSnapshot();
  });

  it("shows sign In option if user is not authenticated", () => {
    const menuText = "Sign In";

    const { getByText } = render(<AuthUserMenuItem />);

    expect(getByText(menuText))
  });

  it("shows login option if user is not authenticated", () => {
    const menuText = "Log out";
    const authUser_ : AuthenticatedUser = {
      id: "1",
      email: "abc@gmail.com",
      displayName: "abc",
    };

    (useAuth as jest.Mock).mockReturnValue({
      isAuthenticated: true,
      authUser: authUser_,
    });

    const { getByText } = render(<AuthUserMenuItem />);

    expect(getByText(menuText))
    expect(getByText(`Hi ${authUser_.email}`));
  });
});
