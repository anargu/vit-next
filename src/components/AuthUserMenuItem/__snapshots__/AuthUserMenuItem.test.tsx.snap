// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthUserMenuItem renders AuthUserMenuItem 1`] = `
<div>
  <div
    class=" relative text-right"
  >
    <div>
      <span
        class="cursor-pointer inline-flex items-center gap-2"
      >
        <svg />
      </span>
    </div>
    <div
      class="hidden right-0
        absolute z-30 min-w-[96px] text-base list-none bg-white rounded divide-y
        divide-gray-100 shadow dark:bg-gray-700 dark:divide-gray-600"
    >
      <ul
        aria-labelledby="dropdown"
        class="py-1"
      >
        <li>
          <span
            class="block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
          >
            Sign In
          </span>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
